maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,492 @@
+
+            if type(action) == PlaceAction:
+                pm_id = action.pm_id
+                vm: VirtualMachine = self._pending_vm_request_payload[vm_id].vm_info
+                lifetime = vm.lifetime
+
+                # Update VM information.
+                vm.pm_id = pm_id
+                vm.creation_tick = cur_tick
+                vm.deletion_tick = cur_tick + lifetime
+                vm.cpu_utilization = vm.get_utilization(cur_tick=cur_tick)
+
+                # Pop out the VM from pending requests and add to live VM dict.
+                self._pending_vm_request_payload.pop(vm_id)
+                self._live_vms[vm_id] = vm
+
+                # TODO: Current logic can not fulfill the oversubscription case.
+
+                # Update PM resources requested by VM.
+                pm = self._machines[pm_id]
+                pm.place_vm(vm.id)
+                pm.cpu_cores_allocated += vm.cpu_cores_requirement
+                pm.memory_allocated += vm.memory_requirement
+                # Calculate the PM's utilization.
+                pm_cpu_utilization = (
+                    (pm.cpu_cores_capacity * pm.cpu_utilization + vm.cpu_cores_requirement * vm.cpu_utilization)
+                    / pm.cpu_cores_capacity
+                )
+                pm.set_cpu_utilization(cpu_utilization=pm_cpu_utilization)
+                pm.energy_consumption = self._cpu_utilization_to_energy_consumption(cpu_utilization=pm.cpu_utilization)
+                self._successful_placement += 1
+            elif type(action) == PostponeAction:
+                postpone_frequency = action.postpone_frequency
+                remaining_buffer_time = self._pending_vm_request_payload[vm_id].remaining_buffer_time
+                # Either postpone the requirement event or failed.
+                self._postpone_vm_request(
+                    postpone_type=PostponeType.Agent,
+                    vm_id=vm_id,
+                    remaining_buffer_time=remaining_buffer_time - postpone_frequency * self._config["delay_duration"]
+                )
+
+    def _download_processed_data(self):
+        """Build processed data."""
+
+        data_root = StaticParameter.data_root
+        build_folder = os.path.join(data_root, self._scenario_name, ".build", self._topology)
+
+        source = self._config["PROCESSED_DATA_URL"]
+        download_file_name = source.split('/')[-1]
+        download_file_path = os.path.join(build_folder, download_file_name)
+
+        # Download file from the Azure blob storage.
+        if not os.path.exists(download_file_path):
+            logger.info_green(f"Downloading data from {source} to {download_file_path}.")
+            download_file(source=source, destination=download_file_path)
+        else:
+            logger.info_green("File already exists, skipping download.")
+
+        logger.info_green(f"Unzip {download_file_path} to {build_folder}")
+        # Unzip files.
```

Comment:
how about check the final file not the zipped file, what's the file size difference between the ori file and the zipped file?

Reply:
``` python
       if (not os.path.exists(vm_table_data_path)) or (not os.path.exists(cpu_readings_data_path)):
            self._download_processed_data()
```
So in line 126, have already checked if the files (final files) exist. 
If not, then it will go to this function to download the files. 

Hence, here, is going to check if the zipped file exist or not, if not, then it will trigger the download pipeline, and unzip it.